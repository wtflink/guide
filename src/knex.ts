import { knex as _knex } from 'knex';

import config from './knexfile';

const knex = _knex(config);

export default knex;
