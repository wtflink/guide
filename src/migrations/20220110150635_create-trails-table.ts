import { Knex } from 'knex';

export async function up(knex: Knex): Promise<any> {
  return knex.schema.createTable('trails', (table: Knex.TableBuilder) => {
    table.text('id').primary();
    table.text('url').nullable();
    table.timestamp('created_at').defaultTo(knex.fn.now());
    table.timestamp('expire_at');
  });
}

export async function down(knex: Knex): Promise<any> {
  return knex.schema.dropTable('trails');
}
