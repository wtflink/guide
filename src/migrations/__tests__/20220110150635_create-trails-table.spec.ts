import mockKnex from 'mock-knex';

import knex from '../../knex';

import { up, down } from '../20220110150635_create-trails-table';

let tracker: mockKnex.Tracker;

beforeAll(() => {
  mockKnex.mock(knex);
  tracker = mockKnex.getTracker();
});

beforeEach(() => {
  tracker.install();
});

afterEach(() => {
  tracker.uninstall();
});

function setupMockQuery({ responses = {} }: Record<string, any> = {}) {
  const sql: string[] = [];

  tracker.on('query', (query) => {
    let boundSQL = query.sql;

    if (query.bindings) {
      query.bindings.forEach((_: any, i: number) => {
        boundSQL = boundSQL.replace(`$${i + 1}`, '?');
      });
    }

    const rawQuery = knex.raw(boundSQL, query.bindings).toString();

    sql.push(rawQuery);

    const response = responses[rawQuery];

    query.response(response || []);
  });

  return {
    getSQL: () => sql.join(` \n`),
  };
}

describe('#up', () => {
  it('should match SQL snapshot', async () => {
    const { getSQL } = setupMockQuery();

    await up(knex);

    expect(getSQL()).toMatchSnapshot();
  });
});

describe('#down', () => {
  it('should match SQL snapshot', async () => {
    const { getSQL } = setupMockQuery();

    await down(knex);

    expect(getSQL()).toMatchSnapshot();
  });
});
