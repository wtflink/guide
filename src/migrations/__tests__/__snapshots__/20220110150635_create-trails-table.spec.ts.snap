// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#down should match SQL snapshot 1`] = `"drop table \\"trails\\""`;

exports[`#up should match SQL snapshot 1`] = `
"create table \\"trails\\" (\\"id\\" text, \\"url\\" text null, \\"created_at\\" timestamptz default CURRENT_TIMESTAMP, \\"expire_at\\" timestamptz) 
alter table \\"trails\\" add constraint \\"trails_pkey\\" primary key (\\"id\\")"
`;
